src/common/primitive_serialization.hpp

```diff
@@ -0,0 +1,76 @@
+/*******************************************************************************
+* Copyright 2021-2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#ifndef COMMON_PRIMITIVE_SERIALIZATION_HPP
+#define COMMON_PRIMITIVE_SERIALIZATION_HPP
+
+#include "common/c_types_map.hpp"
+#include "common/primitive_attr.hpp"
+#include "common/serialization.hpp"
+#include "common/type_helpers.hpp"
+#include "oneapi/dnnl/dnnl.h"
+
+namespace dnnl {
+namespace impl {
+namespace primitive_serialization {
+
+void serialize_post_ops(
+        serialization_stream_t &sstream, const post_ops_t &post_ops);
```

Comment:
This code seems excessively verbose and the style is unaligned with how the GPU serialization tends to work. What if we just use a templated function instead?

```
namespace  primitive {

template <typename T> void serialize(serialization_stream_t sstream, const T &t);

}
```
And then use `extern template` in this file. After that, usage is `primitive::serialize()` which seems more readable. Debatably, we could also drop the namespace guard, as it doesn't seem like it serves much of a purpose as implementing multiple serializers for these core types seems like it would be an antipattern.

Reply:
I agree, switching to overloading + dropping the namespace look reasonable to me.

+ @densamoilov who implemented this originally.

Reply:
Dropped the namespace and switched to overloads (we don't need template functions yet so overloads work fine).